From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: [PATCH] zfcp: optimize zfcp_qdio_account
References: bnc#532063,LTC#55526

From: Heiko Carstens <heiko.carstens@de.ibm.com>

Remove expensive ktime_get()/ktime_us_delta() functions from the hot
path and use get_clock_monotonic() instead.
This elimates seven function calls and avoids a lot of unnecessary
calculations.

Signed-off-by: Heiko Carstens <heiko.carstens@de.ibm.com>

Acked-by: John Jolly <jjolly@suse.de>
---

 drivers/s390/scsi/zfcp_def.h  |    2 +-
 drivers/s390/scsi/zfcp_qdio.c |   11 +++++------
 2 files changed, 6 insertions(+), 7 deletions(-)

Index: linux-sles11/drivers/s390/scsi/zfcp_def.h
===================================================================
--- linux-sles11.orig/drivers/s390/scsi/zfcp_def.h	2009-08-17 15:31:27.000000000 +0200
+++ linux-sles11/drivers/s390/scsi/zfcp_def.h	2009-08-17 15:31:29.000000000 +0200
@@ -472,7 +472,7 @@
 	spinlock_t		req_q_lock;	   /* for operations on queue */
 	int			req_q_pci_batch;   /* SBALs since PCI indication
 						      was last set */
-	ktime_t			req_q_time; /* time of last fill level change */
+	unsigned long long	req_q_time; /* time of last fill level change */
 	u64			req_q_util; /* for accounting */
 	spinlock_t		qdio_stat_lock;
 	u32			fsf_req_seq_no;	   /* FSF cmnd seq number */
Index: linux-sles11/drivers/s390/scsi/zfcp_qdio.c
===================================================================
--- linux-sles11.orig/drivers/s390/scsi/zfcp_qdio.c	2009-08-17 15:31:27.000000000 +0200
+++ linux-sles11/drivers/s390/scsi/zfcp_qdio.c	2009-08-17 15:31:29.000000000 +0200
@@ -77,16 +77,15 @@
 }
 
 /* this needs to be called prior to updating the queue fill level */
-static void zfcp_qdio_account(struct zfcp_adapter *adapter)
+static inline void zfcp_qdio_account(struct zfcp_adapter *adapter)
 {
-	ktime_t now;
-	s64 span;
+	unsigned long long now, span;
 	int free, used;
 
 	spin_lock(&adapter->qdio_stat_lock);
-	now = ktime_get();
-	span = ktime_us_delta(now, adapter->req_q_time);
-	free = max(0, atomic_read(&adapter->req_q.count));
+	now = get_clock_monotonic();
+	span = (now - adapter->req_q_time) >> 12;
+	free = atomic_read(&adapter->req_q.count);
 	used = QDIO_MAX_BUFFERS_PER_Q - free;
 	adapter->req_q_util += used * span;
 	adapter->req_q_time = now;
